sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala

```diff
@@ -315,4 +315,65 @@ case class StringLength(child: Expression) extends UnaryExpression with ExpectsI
   }
 }
 
+/**
+ * Like ConcatWS below, but taking an array of strings.
+ */
+case class ConcatWS(sep: Expression, child: Expression)
+  extends Expression with ExpectsInputTypes {
+  override def dataType: DataType = StringType
+  override def nullable: Boolean = true
+  override def foldable: Boolean = child.foldable && sep.foldable
+  override def children: Seq[Expression] = sep :: child :: Nil
+  override def expectedChildTypes: Seq[DataType] = StringType :: ArrayType(StringType) :: Nil
```

Comment:
This probably not right, we support both `StringType` and `ArrayType(StringType)`, and maybe we'd better not inherit from the trait `ExpectsInputTypes`